src/super_gradients/training/utils/callbacks/callbacks.py

```diff
@@ -948,3 +954,206 @@ def create_lr_scheduler_callback(
         raise ValueError(f"Unknown lr_mode: {lr_mode}")
 
     return sg_lr_callback
+
+
+class ExtremeBatchCaseVisualizationCallback(Callback, ABC):
+    """
+    ExtremeBatchCaseVisualizationCallback
+
+    A base class for visualizing worst/best validation batches in an epoch
+     according to some metric or loss value, with Full DDP support.
+
+    Images are saved with training_hyperparams.sg_logger.
+
+    :param metric_name: str,will be the metric which the model checkpoint will be saved according to, and can be set to any
+      of the following:
+
+        a metric name (str) of one of the metric objects from the training_hyperparams.valid_metrics_list
+
+        a "component_name" if some metric in valid_metrics_list has an attribute component_names. In such cas it
+        is a list referring to the names of each entry in the output metric (torch tensor of size n).
+
+        one of "loss_logging_items_names" i.e which will correspond to an item returned during the
+        loss function's forward pass (see loss docs in Trainer.train(..)).
+
+    :param max: bool, Whether to take the batch corresponding to the max value of the metric/loss or
+     the minimum (default=False).
+
+    :param freq: int, epoch frequency to perform all of the above (default=1).
+
+     Inheritors should implement process_extreme_batch which returns an image, as an np.array (uint8) with shape BCHW.
+    """
+
+    def __init__(self, metric_name: str, max: bool = False, freq: int = 1):
+        self.metric_name = metric_name
+        self.metric = None
+        self.max = max
+        self.freq = freq
+        self.extreme_score = -1 * np.inf if max else np.inf
+
+        self.extreme_batch = None
+        self.extreme_preds = None
+        self.extreme_targets = None
+
+        self._first_call = True
+        self._idx_loss_tuple = None
+        self._tag = f"max_{self.metric_name}_batch" if self.max else f"min_{self.metric_name}_batch"
+
+        super(ExtremeBatchCaseVisualizationCallback, self).__init__()
+
+    @abstractmethod
+    def process_extreme_batch(self) -> np.array:
+        raise NotImplementedError
+
+    def on_training_start(self, context: PhaseContext) -> None:
+        """
+        On train start we set the metric (if the metric_name does not corresponf to a loss).
+        :param context: Phase context
+        :return:
+        """
+        if not hasattr(context.valid_metrics, self.metric_name):
```

Comment:
I feel that logic of juggling with metrics here is unnecessary complicated. SRP and DRY are just crying in corner. I was not able to wrap my head around after reading this code first time. I was not successful second time either.
To me it should be done in a completely different manner which I believe would end up with a much more clearer implementation.
Here's what I suggest.

1) Introduce a concept of "batch metrics", which in the most simple case is a dictionary of `{metric_name1: metric value, metric_name2: metric value2, ...}`.
This concept is to be managed by a Trainer class and available inside the `context`. So all the per-batch metrics (Which is loss values and metrics that we can compute per-batch) should be available through this interface.
Pros: In the callback class now have only the simple access to metric:
```
if _is_more_extreme(context.batch_metrics[self.monitored_value]):
  self.extreme_score = score
  self.extreme_batch = context.inputs
  self.extreme_preds = context.preds
  self.extreme_targets = context.target
```

2) For IoU metric all we have to do is to ensure it is also computed per-batch and stored in `batch metrics` container. And that's it.

I feel this change would greatly improve the responsibility boundaries between classes and reduce the amount of duplicated code.

Reply:
I did think of it, but I feel a bit worried about interfering with metrics inside the MetricCollection object, therefore I rather have it somehere else.

From what I understand, you suggest passing a dictionary that would suppply the information on what metrics we would like to compute per batch, correct?

There is a drawback here if so, since you are asking the user to pass additional info which IMO sould be coupled with this callback.

Reply:
Passing metric explicitly now.